beacon-chain/state/state-native/validator_index_cache.go

```diff
@@ -0,0 +1,97 @@
+package state_native
+
+import (
+	"sync"
+
+	"github.com/pkg/errors"
+	"github.com/prysmaticlabs/prysm/v5/config/features"
+	fieldparams "github.com/prysmaticlabs/prysm/v5/config/fieldparams"
+	consensus_types "github.com/prysmaticlabs/prysm/v5/consensus-types"
+	"github.com/prysmaticlabs/prysm/v5/consensus-types/primitives"
+	"github.com/prysmaticlabs/prysm/v5/encoding/bytesutil"
+	log "github.com/sirupsen/logrus"
+)
+
+// validatorIndexCache maintains a mapping from validator public keys to their indices within the beacon state.
+// It includes a lastFinalizedIndex to track updates up to the last finalized validator index,
+// and uses a mutex for concurrent read/write access to the cache.
+type validatorIndexCache struct {
+	indexMap           map[[fieldparams.BLSPubkeyLength]byte]primitives.ValidatorIndex // Maps BLS public keys to validator indices.
+	lastFinalizedIndex int                                                             // Index of the last finalized validator to avoid reading validators before this point.
+	mutex              sync.RWMutex
+}
+
+// newValidatorIndexCache initializes a new validator index cache with an empty index map.
+func newValidatorIndexCache() *validatorIndexCache {
+	return &validatorIndexCache{
+		indexMap: make(map[[fieldparams.BLSPubkeyLength]byte]primitives.ValidatorIndex),
+	}
+}
+
+// getValidatorIndex retrieves the validator index for a given public key from the cache.
+// If the public key is not found in the cache, it searches through the state starting from the last finalized index.
+func (b *BeaconState) getValidatorIndex(pubKey [fieldparams.BLSPubkeyLength]byte) (primitives.ValidatorIndex, bool) {
+	b.validatorIndexCache.mutex.RLock()
+	index, found := b.validatorIndexCache.indexMap[pubKey]
+	b.validatorIndexCache.mutex.RUnlock()
+	if found {
+		return index, true
+	}
+
+	finalizedIndex := b.validatorIndexCache.lastFinalizedIndex
+	pubKeys, err := b.publicKeysSinceIndexReadOnly(finalizedIndex)
```

Comment:
This will also return the last finalized validator, which is not a bug but also not exactly what we want. The "correct" thing to do is to call the function with `finalizedIndex + 1` and later write `index := primitives.ValidatorIndex(finalizedIndex + i + 1)`. Similarly for `saveValidatorIndices`.